File: Lexer.cs
using System;
using System.Collections.Generic;

namespace TinyLang {
	enum TokenKind {
		End, Error,

		String, Int, Float, Boolean,

		Plus, Minus, Star, Slash, At,

		Let, Var, Identifier, Struct, Function,
		Return, If, Else, While, Do, Auto,
		Dot, Comma, Colon, SemiColon, Equals,
		OpenParen, CloseParen, OpenCurly, CloseCurly,
		OpenSquare, CloseSquare, New,

		NotEqual, EqualEqual, Greater, Less, GreaterEqual, LessEqual,
	}
	
	class Token {
		public readonly string Lexeme;
		public readonly TokenKind Kind;
		public readonly int Column, Line;

		public Token(string lexeme, TokenKind kind, int column, int line) {
			this.Lexeme = lexeme;
			this.Kind = kind;
			this.Column = column;
			this.Line = line;
		}
	}

	class Lexer {
		readonly string source;
		int ip, line, column;

		static Dictionary<string, TokenKind> KeyWords = new Dictionary<string, TokenKind>() {
			{ "var", TokenKind.Var },
			{ "let", TokenKind.Let },
			{ "auto", TokenKind.Auto },
			{ "return", TokenKind.Return },
			{ "fn", TokenKind.Function },
			{ "new", TokenKind.New },
			{ "struct", TokenKind.Struct },
			{ "do", TokenKind.Do },
			{ "while", TokenKind.While },
			{ "if", TokenKind.If },
			{ "else", TokenKind.Else },
			{ "true", TokenKind.Boolean },
			{ "false", TokenKind.Boolean },
		};

		static Dictionary<char, TokenKind> Singles = new Dictionary<char, TokenKind>() {
			{ '+', TokenKind.Plus },
			{ '-', TokenKind.Minus },
			{ '*', TokenKind.Star },
			{ '/', TokenKind.Slash },
			{ '@', TokenKind.At },
			{ '.', TokenKind.Dot },
			{ ',', TokenKind.Comma },
			{ ':', TokenKind.Colon },
			{ ';', TokenKind.SemiColon },
			{ '=', TokenKind.Equals },
			{ '(', TokenKind.OpenParen },
			{ ')', TokenKind.CloseParen },
			{ '{', TokenKind.OpenCurly },
			{ '}', TokenKind.CloseCurly },
			{ '[', TokenKind.OpenSquare },
			{ ']', TokenKind.CloseSquare },
			{ '>', TokenKind.Greater },
			{ '<', TokenKind.Less },
		};

		public Lexer(string source) {
			this.source = source;
			this.line = 1;
			this.column = 1;
		}

		public Token Next() {
			SkipWhitespace();

			if (ip >= source.Length) {
				return new Token("EOF", TokenKind.End, column, line);
			}

			if (IsIdentifier(source[ip])) {
				return Identifier();
			}

			if (Char.IsDigit(source[ip])) {
				return Number();
			}

			if (source[ip] == '"') {
				return String();
			}

			if (Peek() == '!' && Peek(1) == '=') {
				return MakeDouble(TokenKind.NotEqual);
			}

			if (Peek() == '=' && Peek(1) == '=') {
				return MakeDouble(TokenKind.EqualEqual);
			}

			if (Peek() == '>' && Peek(1) == '=') {
				return MakeDouble(TokenKind.GreaterEqual);
			}

			if (Peek() == '<' && Peek(1) == '=') {
				return MakeDouble(TokenKind.LessEqual);
			}
			
			if (Singles.ContainsKey(source[ip])) {
				Token t = new Token(source[ip].ToString(), Singles[source[ip]], column, line);
				Advance();
				return t;
			}

			return new Token("ERROR", TokenKind.End, column, line);
		}

		Token MakeDouble(TokenKind kind) {
			Advance();
			Advance();
			return new Token(source[(ip - 2)..ip], kind, column - 2, line);
		}

		Token Identifier() {
			int start = ip;
			int col = column;
			TokenKind kind = TokenKind.Identifier;

			while (ip < source.Length && (IsIdentifier(source[ip]) || char.IsDigit(source[ip]))) {
				Advance();
			}

			string lexeme = source[start..ip];

			if (KeyWords.ContainsKey(lexeme)) {
				kind = KeyWords[lexeme];
			}

			return new Token(lexeme, kind, col, line);
		}

		Token String() {
			Advance();

			int start = ip;
			int col = column;

			while (ip < source.Length && source[ip] != '"') {
				Advance();

				if (ip >= source.Length) {
					throw new Exception("Unterminated string");
				}
			}
			Advance();

			return new Token(source[start..(ip-1)], TokenKind.String, col, line);
		}

		Token Number() {
			int start = ip;
			int col = column;
			TokenKind kind = TokenKind.Int;

			while (ip < source.Length && Char.IsDigit(source[ip])) {
				Advance();

				if (ip < source.Length && source[ip] == '.') {
					if (kind == TokenKind.Float) {
						throw new Exception("Floating point number contains multiple decimals");
					}

					kind = TokenKind.Float;
					Advance();
				}
			}

			return new Token(source[start..ip], kind, col, line);
		}

		bool IsIdentifier(char ch) {
			return Char.IsLetter(ch) || ch == '_';
		}

		void Advance() {
			ip++;
			column++;
		}

		char Peek(int depth = 0) {
			if (ip + depth >= source.Length) return '\0';
			return source[ip + depth];
		}

		void SkipWhitespace() {
			while (ip < source.Length) {
				switch(source[ip]) {
					case '#': {
						while (ip < source.Length && source[ip] != '\n') {
							Advance();
						}
						break;
					}

					case '\n': {
						line++;
						ip++;
						column = 1;
						break;
					}

					case ' ':
					case '\t':
					case '\b':
					case '\r':
						Advance();
						break;
					
					default:
						return;
				}
			}
		}
	}
}

File: lexer/lexer.go
package lexer

import "fmt"

type Lexer struct {
	source            string
	line, column, pos int
}

func New(source string) *Lexer {
	return &Lexer{source, 1, 1, 0}
}

func (lexer *Lexer) Next() *Token {
	lexer.skipWhitespace()

	if lexer.isAtEnd() {
		return lexer.makeEof()
	}

	// Identifiers cannot start with digits
	if isAlpha(lexer.peek()) {
		return lexer.readIdentifier()
	}

	if isDigit(lexer.peek()) {
		return lexer.readDigit()
	}

	if lexer.peek() == '"' {
		return lexer.readString()
	}

	// This is a fallthrough, which will return an error *Token otherwise
	return lexer.readChars()
}

// --- Private ---
func (lexer *Lexer) makeEof() *Token {
	return &Token{EOF, "EndOfFile", lexer.line, lexer.column}
}

func (lexer *Lexer) makeError(msg string, arg ...any) *Token {
	return &Token{ERROR, fmt.Sprintf(msg, arg...), lexer.line, lexer.column}
}

func (lexer *Lexer) makeToken(kind TokenKind, lexeme string, column int) *Token {
	return &Token{kind, lexeme, lexer.line, column}
}

func (lexer *Lexer) peek() byte {
	if lexer.pos >= len(lexer.source) {
		return 0
	}
	return lexer.source[lexer.pos]
}

func (lexer *Lexer) advance() {
	lexer.column++
	lexer.pos++
}

func (lexer *Lexer) isAtEnd() bool {
	return lexer.pos >= len(lexer.source)
}

func isDigit(char byte) bool {
	return char >= '0' && char <= '9'
}

func isAlpha(char byte) bool {
	return char >= 'a' && char <= 'z' || char >= 'A' && char <= 'Z' || char == '_'
}

func isIdentifier(char byte) bool {
	return isAlpha(char) || isDigit(char)
}

func getKeyword(lexeme string) TokenKind {
	kind, ok := KeyWords[lexeme]

	if !ok {
		return IDENTIFIER
	} else {
		return kind
	}
}

func (lexer *Lexer) match(expected byte) bool {
	if lexer.isAtEnd() {
		return false
	}
	if lexer.peek() != expected {
		return false
	}

	lexer.advance()
	return true
}

func (lexer *Lexer) skipWhitespace() {
	for !lexer.isAtEnd() {
		switch lexer.peek() {
		case '#':
			{
				for !lexer.isAtEnd() && lexer.peek() != '\n' {
					lexer.advance()
				}
			}

		case '\n':
			lexer.line++
			lexer.column = 1
			lexer.pos++

		case ' ', '\t', '\b', '\r':
			lexer.advance()

		default:
			return
		}
	}
}

func (lexer *Lexer) readChars() *Token {
	current := lexer.peek()
	lexer.advance()

	size := 1
	kind := ERROR

	switch current {
	case '+':
		if lexer.match('=') {
			kind = PLUS_EQUAL
			size = 2
			break
		}
		kind = PLUS
	case '-':
		if lexer.match('=') {
			kind = MINUS_EQUAL
			size = 2
			break
		}
		kind = MINUS
	case '*':
		if lexer.match('=') {
			kind = STAR_EQUAL
			size = 2
			break
		}
		kind = STAR
	case '/':
		if lexer.match('=') {
			kind = SLASH_EQUAL
			size = 2
			break
		}
		kind = SLASH
	case '(':
		kind = OPENPAREN
	case ')':
		kind = CLOSEPAREN
	case '{':
		kind = OPENCURLY
	case '}':
		kind = CLOSECURLY
	case '[':
		kind = OPENSQUARE
	case ']':
		kind = CLOSESQUARE
	case ':':
		kind = COLON
	case ';':
		kind = SEMICOLON
	case '.':
		kind = DOT
	case ',':
		kind = COMMA

	case '&':
		if lexer.match('&') {
			kind = AND
			size = 2
			break
		}
		kind = AMPERSAND

	case '|':
		if lexer.match('|') {
			kind = OR
			size = 2
			break
		}
		kind = PIPE

	case '=':
		if lexer.match('=') {
			kind = EQUAL_EQUAL
			size = 2
			break
		} else if lexer.match('>') {
			kind = FAT_ARROW
			size = 2
			break
		}
		kind = EQUAL

	case '!':
		if lexer.match('=') {
			kind = NOT_EQUAL
			size = 2
			break
		}
		kind = BANG

	case '>':
		if lexer.match('=') {
			kind = GREATER_EQUAL
			size = 2
			break
		}
		kind = GREATER

	case '<':
		if lexer.match('=') {
			kind = LESS_EQUAL
			size = 2
			break
		}
		kind = LESS

	default:
		return lexer.makeError("Unknown character found '%q'", string(current))
	}

	return lexer.makeToken(kind, lexer.source[lexer.pos-size:lexer.pos], lexer.column-size)
}

func (lexer *Lexer) readDigit() *Token {
	start := lexer.pos
	start_col := lexer.column
	kind := INT

	// TODO: Allow floats
	for !lexer.isAtEnd() && isDigit(lexer.peek()) {
		lexer.advance()

		if lexer.peek() == '.' {
			if kind == FLOAT {
				return lexer.makeError("Floating point number cannot have multiple decimals %d:%d", lexer.line, lexer.column)
			}

			kind = FLOAT
			lexer.advance()
		}
	}

	return lexer.makeToken(kind, lexer.source[start:lexer.pos], start_col)
}

func (lexer *Lexer) readString() *Token {
	lexer.advance()

	start := lexer.pos
	start_col := lexer.column

	for !lexer.isAtEnd() && lexer.peek() != '"' {
		lexer.advance()
	}

	lexer.advance()
	return lexer.makeToken(STRING, lexer.source[start:lexer.pos-1], start_col)
}

func (lexer *Lexer) readIdentifier() *Token {
	start := lexer.pos
	start_col := lexer.column

	lexer.advance()

	for !lexer.isAtEnd() && isIdentifier(lexer.peek()) {
		lexer.advance()
	}

	lexeme := lexer.source[start:lexer.pos]
	return lexer.makeToken(getKeyword(lexeme), lexeme, start_col)
}
